### PR TITLE
Implement binding docs.

### PR DESCRIPTION
**Pull request by MichaelHatherly**
This splits out the bindings-related commits from #12088, which are needed to avoid documenting the values of constants rather than the constants themselves, ie.

```
julia> "a"
       a = 1

julia> "b"
       b = 1

julia> Main.__META__
ObjectIdDict with 2 entries:
  1                        => Base.Markdown.MD(Any[Base.Markdow…
  #= circular reference =# => Base.Markdown.MD(Any[Base.…
```

cc @one-more-minute: I'm not quite sure about the lines https://github.com/JuliaLang/julia/pull/12088/files#diff-5a9ad42cbc2e070b47b9d63118501e2fR330 and https://github.com/JuliaLang/julia/pull/12088/files#diff-5a9ad42cbc2e070b47b9d63118501e2fR296 where `@var` syntax is supported. What's the intention behind that?

### CODE REVIEW

**Comment by MichaelHatherly** [comment]
Thanks @yuyichao, seems my title was also a work in progress.

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -297,22 +350,23 @@ function docm(meta, def, define = true)
...
     isexpr(def′, :abstract)    && return namedoc(meta, def, namify(def′))
     isexpr(def′, :bitstype)    && return namedoc(meta, def, def′.args[2])
+    isexpr(def′, :typealias)   && return  vardoc(meta, def, namify(def′))
     isexpr(def′, :module)      && return namedoc(meta, def, def′.args[2])
-    isexpr(def′, :(=), :const) && return namedoc(meta, def, namify(def′))
+    isexpr(def′, :(=), :const) && return  vardoc(meta, def, namify(def′))
 
-    isexpr(def′, :macrocall) && (def = namify(def′))
-    return objdoc(meta, def)
+    objdoc(meta, def′)
```
Comment (one-more-minute) [inline base/docs/Docs.jl]: Just making sure I understand – why the change to def-prime here?

File: base/docs/Docs.jl [hunk 2]
```diff
@@ -320,7 +374,7 @@ function docm(ex)
     isexpr(ex, :->) && return docm(ex.args...)
     isexpr(ex, :call) && return findmethod(ex)
     isexpr(ex, :macrocall) && (ex = namify(ex))
-    :(doc($(esc(ex))))
+    :(doc(@var($(esc(ex)))))
```
Comment (one-more-minute) [inline base/docs/Docs.jl]: I think it makes sense to explicitly check for valid `@var` syntax here and fall back as before, otherwise you essentially disallow retreiving documentation for values.

File: base/docs/Docs.jl (hunk 1 above)
Comment (MichaelHatherly) [inline base/docs/Docs.jl]: As an example, documenting a module that is defined during bootstrap with

```
"..."
MyModule
```

takes that code path and we end up documenting ``nothing`` if ``def`` is used instead of ``def′``.

File: base/docs/Docs.jl (hunk 2 above)
Comment (MichaelHatherly) [inline base/docs/Docs.jl]: > explicitly check for valid ``@var`` syntax

Sure, will do.

File: base/docs/Docs.jl (hunk 1 above)
Comment (one-more-minute) [inline base/docs/Docs.jl]: Hmm, wouldn't `objdoc` then end up redefining the module (or re-running `def′` in general) in that case?

(It doesn't matter all that much for the sake of this PR, since I'm going to get rid of that fallback and redo module documentation anyway)

Comment (MichaelHatherly) [inline base/docs/Docs.jl]: It's only for a module name rather that an expression ``module MyModule ... end``. Those are caught by the ``isexpr(def′, :module) && ...`` on `366`.

All the cases of expressions that could be redefined are covered by the block above, I believe.

**Comment by MichaelHatherly** [comment]
Thanks.
